platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java

```diff
@@ -13,204 +13,259 @@
+   * denoted using 2-tuple combinations of variables: [felixFileFromCache, felixFileFromConfig]
    *
-   * <p>[null, X] (2) Unrecognized config with a felix file prop to track. The config's presence in
-   * etc was previously not being tracked by the system and we should start tracking it.
+   * <p>[null, null] (1) Config was created from config admin directly; not by dropping a file.
+   * There is nothing to track.
    *
-   * <p>[Y, null] [Y, X] (3) Felix prop was externally changed or removed. Revert this change.
+   * <p>[null, X] (2) Some config was just created by dropping a file into etc; track the file.
+   * Update the properties on disk if necessary and perform any special first time processing.
    *
-   * <p>[Y, Y] (4) Felix file prop did not change, do nothing. This is the happy path.
+   * <p>[Y, null] [Y, X] (3) Felix prop was externally changed or removed. This is an error.
    *
-   * @param event the configuration event for which an update is needed
+   * <p>[Y, Y] (4) Felix file prop did not change, do nothing. This is the happy path. Properties
+   * will be updated on disk if necessary.
    */
-  private void handleUpdate(ConfigurationEvent event) {
-    Configuration configuration = getConfiguration(event);
-    if (configuration == null) {
-      LOGGER.debug("Configuration was null for pid {}", event.getPid());
-      return;
-    }
+  @Override
+  public void handleStore(ConfigurationContext context) throws IOException {
+    String pid = context.getServicePid();
+    File fileFromConfig = context.getConfigFile();
 
-    FelixConfig felixConfig = new FelixConfig(configuration);
-    String pid = felixConfig.getPid();
-    File felixFileNew = felixConfig.getFelixFile();
-    File felixFileOriginal = pidFileMap.get(pid);
+    CachedConfigData cachedConfigData = pidDataMap.get(pid);
+    File fileFromCache = (cachedConfigData != null) ? cachedConfigData.getFelixFile() : null;
 
-    // Config not tracked and will not be tracked? (1)
-    if (felixFileOriginal == null && felixFileNew == null) {
+    if (fileFromCache == null && fileFromConfig == null) {
+      // This config doesn't have an etc file, so we ignore this case (1)
       LOGGER.debug("Was not tracked and will not track pid {}", pid);
       return;
     }
 
-    boolean filePropChanged = felixConfig.filePropChanged(felixFileOriginal);
-
-    // Felix file prop did not change? (4)
-    if (filePropChanged) {
-      // Are we seeing the config for the first time? (2) (3)
-      if (felixFileOriginal != null) {
-        try {
-          felixConfig.setFelixFile(felixFileOriginal);
-        } catch (IOException e) {
-          LOGGER.error("Could not set felix file name, error writing to config admin: ", e);
-        }
-        return;
-      }
-      pidFileMap.put(pid, felixFileNew);
+    if (fileFromCache == null) {
+      // An etc config file was just dropped and we're seeing it for the first time (2)
```

Comment:
✏️ These are excellent uses of inline comments to explicate workflow. However, you might consider extracting each block into its own private method. Then this method would be setup; check conditions and comment; call methods for each case.

It's a matter of opinion when each of the conditional blocks become large enough that breaking the method up like that makes sense. My feeling is that it's currently right on the edge and any additional business logic in these blocks would make it very hard to follow.

Reply:
I tried doing just that. But I make heavy use of `return`ing early. The tradeoff is that I can't make elegant use of `return` statements anymore, and will need redundant if-statements in `handleStore` if I want to bail early. The component methods would have to return boolean. 

Also there's a few local variables in play and I think passing a lot of arguments around might get confusing as well. 

Now, regarding case (3) when the felix prop is illegally changed, I could definitely pull that out if it makes reading easier. That is a simple function of two files and throws an exception.